### PR TITLE
add a DTensor test for weight tying

### PR DESCRIPTION
**Pull request by tianyu-l**
Weight tying is useful when we'd like to share weights (and their gradients) between two modules, e.g. the word/token embedding module and the output linear module in language models. This test demonstrates that with DTensor it can be achieved just as with normal tensor, e.g. using `model.fc.weight = model.embedding.weight`.

To test: `python test/distributed/tensor/parallel/test_tp_examples.py -k test_weight_tying`

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116475

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @wconstab @yf225

### CODE REVIEW

File: test/distributed/_tensor/test_dtensor.py [hunk 1]
```diff
@@ -820,6 +820,53 @@ def test_split_tensor_1D(self) -> None:
...
 
+    @with_comms
+    def test_weight_tying(self):
+        class TestModule(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                # Initialize different weights for embedding and fc.
+                torch.manual_seed(1)
+                self.embedding = torch.nn.Embedding(16, 8)
+                torch.manual_seed(2)
```
Comment (XilunWu) [inline test/distributed/_tensor/test_dtensor.py:831]: two questions:
1. is there any reason why we need to manually set seed twice? If we want to control the randomness, we only need to have the first `manual_seed`.
2. why do we have it inside the class constructor `__init__`? This means, every `TestModule` object will have the same embedding and Linear parameter. Is that what we want? If we only want to control the randomness so that every time the test runs the initialization result will be the same, we can put `manual_seed` before the line `model = TestModule().to(self.device_type)` (actually this is even unnecessary because [`MultiProcessTestCase`](https://github.com/pytorch/pytorch/blob/6243dbb5c08e9b3f87bc7dc74816b2ceb6957a94/torch/testing/_internal/common_distributed.py#L961) will call `manual_seed(0)` and `DTensorTestBase` is a subclass of `MultiProcessTestCase`).

File: test/distributed/_tensor/test_dtensor.py [hunk 2]
```diff
@@ -820,6 +820,53 @@ def test_split_tensor_1D(self) -> None:
...
+                self.fc = torch.nn.Linear(8, 16)
+
+            def forward(self, x):
+                return self.fc(self.embedding(x))
+
+        model = TestModule().to(self.device_type)
```
Comment (XilunWu) [inline test/distributed/_tensor/test_dtensor.py:837]: if you want to control the randomness, you can have `manual_seed` right before initializing the module

File: test/distributed/_tensor/test_dtensor.py (hunk 1 above)
Comment (tianyu-l) [reply test/distributed/_tensor/test_dtensor.py:831]: So my point was to make sure two submodules have different weights. Not setting manual seeds twice should also do the work (just by randomness), but mentally it doesn't tell people the two initializations should be different. An alternative I had in mind is to do manual initialization rather than relying on internal gaussian distribution. Maybe I should do that?

File: test/distributed/_tensor/test_dtensor.py [hunk 3]
```diff
@@ -820,6 +820,53 @@ def test_split_tensor_1D(self) -> None:
...
+
+        input_size = [5]
+        torch.manual_seed(0)
+        inp = torch.randint(16, input_size, device=self.device_type)
+
+        # Without weight tying.
+        self.assertNotEqual(
+            model.embedding.weight.to_local(), model.fc.weight.to_local()
+        )
+        output = model(inp)
```
Comment (XilunWu) [inline test/distributed/_tensor/test_dtensor.py:854]: `model` is a distributed module but `inp` is a local `Tensor`. Shouldn't we have `inp` as a `DTensor` (I know the input hook will automatically cast it to `DTensor` but for readability...)

File: test/distributed/_tensor/test_dtensor.py [hunk 4]
```diff
@@ -820,6 +820,53 @@ def test_split_tensor_1D(self) -> None:
...
+            "embedding": ColwiseParallel(),
+            "fc": RowwiseParallel(),
+        }
+        device_mesh = DeviceMesh(self.device_type, list(range(self.world_size)))
+        parallelize_module(model, device_mesh, parallelize_plan)
+        optim = torch.optim.Adam(model.parameters(), lr=0.25)
+
+        input_size = [5]
+        torch.manual_seed(0)
+        inp = torch.randint(16, input_size, device=self.device_type)
```
Comment (XilunWu) [inline test/distributed/_tensor/test_dtensor.py:848]: If you want to create a randomized DTensor, try using [`torch.distributed._tensor.rand()`](https://github.com/pytorch/pytorch/blob/6243dbb5c08e9b3f87bc7dc74816b2ceb6957a94/torch/distributed/_tensor/__init__.py#L216). This function directly creates a DTensor populated with `torch.rand`.

File: test/distributed/_tensor/test_dtensor.py [hunk 5]
```diff
@@ -820,6 +820,53 @@ def test_split_tensor_1D(self) -> None:
...
+        self.assertNotEqual(
+            model.embedding.weight.to_local(), model.fc.weight.to_local()
+        )
+        output = model(inp)
+        output.sum().backward()
+        self.assertNotEqual(
+            model.embedding.weight.grad.to_local(), model.fc.weight.grad.to_local()
+        )
+
+        optim.zero_grad()
```
Comment (XilunWu) [inline test/distributed/_tensor/test_dtensor.py:860]: I see that you want to compare `grad` but what is the usage of `optim` here? This can be done even without `optim`.

**Review by XilunWu — COMMENTED** [review]
overall lgtm! But I have several questions. Also I need @wanchaol 's opinion because I'm not very familiar with weight tying.

File: test/distributed/_tensor/test_dtensor.py [hunk 6]
```diff
@@ -820,6 +820,53 @@ def test_split_tensor_1D(self) -> None:
                 ]
                 assert_array_equal(expected_is_tensor_empty, is_tensor_empty)
 
+    @with_comms
+    def test_weight_tying(self):
```
Comment (wanchaol) [inline test/distributed/_tensor/test_dtensor.py:824]: I think we should add this test case to `tensor/parallel` instead? test_dtensor.py suppose to test some basic functionalities of DTensor that does not involve higher level TP API, but looks like we are using TP API, so let's put this to the TP test folder

File: test/distributed/_tensor/test_dtensor.py (hunk 3 above)
Comment (wanchaol) [reply test/distributed/_tensor/test_dtensor.py:854]: I think this is Tensor Parallel API so it supposes to pass in local tensor instead of DTensor

File: test/distributed/_tensor/test_dtensor.py (hunk 5 above)
Comment (wanchaol) [reply test/distributed/_tensor/test_dtensor.py:860]: yeah we don't need to test the optim update here, we can just test the gradient equality

File: test/distributed/_tensor/test_dtensor.py [hunk 7]
```diff
@@ -820,6 +820,53 @@ def test_split_tensor_1D(self) -> None:
...
+
+        optim.zero_grad()
+
+        # With weight tying.
+        model.fc.weight = model.embedding.weight
+
+        self.assertEqual(model.embedding.weight, model.fc.weight)
+        output = model(inp)
+        output.sum().backward()
+        self.assertEqual(model.embedding.weight.grad, model.fc.weight.grad)
```
Comment (wanchaol) [inline test/distributed/_tensor/test_dtensor.py:868]: can we also assert the grad id actually? to make sure they are really sharing gradient i.e. 
`self.assertEqual(id(model.embedding.weight.grad), id(model.fc.weight.grad))`

File: test/distributed/_tensor/test_dtensor.py (hunk 1 above)
Comment (wanchaol) [reply test/distributed/_tensor/test_dtensor.py:831]: I think this is just to test "weight tying" and a way to make sure embedding/linear have different weights initialized. I am fine with either approach.

**Review by wanchaol — APPROVED** [review]
nice! lgtm, have some comments

File: test/distributed/_tensor/test_dtensor.py (hunk 6 above)
Comment (tianyu-l) [reply test/distributed/_tensor/test_dtensor.py]: Yes I agree. I put it again into `test_tp_examples.py`. Please let me know if you think we'd better create a new test file for this.

**Review by fduwjj — APPROVED** [review]
